objects/src/mock/chain.rs

```diff
@@ -55,3 +55,248 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
 
     chain_mmr
 }
+
+#[cfg(feature = "mock")]
+mod mock {
+    use crate::{
+        assets::{Asset, FungibleAsset, NonFungibleAsset},
+        builder::{
+            AccountBuilder, AccountIdBuilder, FungibleAssetBuilder, NonFungibleAssetBuilder,
+        },
+        Account, AccountError, AccountId, AccountType, AssetError, BlockHeader, ChainMmr, Felt,
+        StorageItem, Vec,
+    };
+    use crypto::{
+        hash::rpo::RpoDigest as Digest,
+        merkle::{DefaultMerkleStore, TieredSmt},
+    };
+    use miden_core::FieldElement;
+    use rand::{Rng, SeedableRng};
+
+    /// Initial timestamp value
+    const TIMESTAMP_START: Felt = Felt::new(1693348223);
+    /// Timestep of timestamp on each new block
+    const TIMESTAMP_STEP: Felt = Felt::new(10);
+
+    #[derive(Default, Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+    pub struct Objects<R: Rng> {
+        accounts: Vec<Account>,
+        fungible_faucets: Vec<(AccountId, FungibleAssetBuilder)>,
+        nonfungible_faucets: Vec<(AccountId, NonFungibleAssetBuilder<R>)>,
+    }
+
+    impl<R: Rng> Objects<R> {
+        pub fn new() -> Self {
+            Self {
+                accounts: vec![],
+                fungible_faucets: vec![],
+                nonfungible_faucets: vec![],
+            }
+        }
+
+        /// Extend this instance with the data of `other`.
+        pub fn extend(&mut self, other: &mut Objects<R>) {
+            self.accounts.extend(other.accounts.drain(..));
+            self.fungible_faucets.extend(other.fungible_faucets.drain(..));
+            self.nonfungible_faucets.extend(other.nonfungible_faucets.drain(..));
+        }
+    }
+
+    /// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+    pub struct MockChain<R: Rng + SeedableRng> {
+        /// An append-only structure used to represent the history of blocks produced for this chain.
+        chain: ChainMmr,
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
+
+        /// History of produced blocks.
+        blocks: Vec<BlockHeader>,
+
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifier_db: TieredSmt,
+
+        // RNG used to seed builders.
+        //
+        // This is used to seed the [AccountBuilder] and the [NonFungibleAssetBuilder].
+        rng: R,
```

Comment:
question: how is `rng` going to be serialized? (mostly just curious)

Reply:
The RNGs derive serde's traits:

https://github.com/rust-random/rand/blob/f3dd0b885c4597b9617ca79987a0dd899ab29fcb/src/rngs/xoshiro128plusplus.rs#L23-L26